### ISSUE TITLE
CUB200-2011 CONTRASTIVE LOSS MAP@R with and without offline miner

### ISSUE THREAD

**Issue by alpoler** [KevinMusgrave/powerful-benchmarker#65]
Hi, ı delete and reinstall library again. Start default experiment with CUB200-2011 dataset and contrastive loss. It founds pretrained MAP@R accuracy as 42. Does anyone happen this problem ? @KevinMusgrave  please help . 
I add only this parameter --experiment_name CONTRA_CUB --dataset {CUB200: {download: False}} for run.py
I work to debug due to fact that i modify code to adapt my offline miner but ı realized that default experiment gives high map@r than that of paper.
![offline_miner](https://user-images.githubusercontent.com/50836811/93600203-edb39980-f9c7-11ea-8e8c-0c29820e0c9a.png)

My offline miner
```
class OfflineMiner_RN(Sampler):

    def __init__(self, margin, embedding, labels, ref_emb, ref_labels, type_of_triplets='semihard', furthest=True,
                 random=True):
        self.margin = margin
        self.type_of_triplets = type_of_triplets
        self.furthest = furthest
        self.Anchor = list(range(labels.shape[0]))
        if random:
            self.Negative = self.RandomIndicies(labels, ref_labels)
            self.Positive = self.PositiveMiner(embedding, labels, ref_emb, ref_labels, type_of_triplets)
        else:
            self.Negative = self.FurtherNearestIndicies(embedding, labels, ref_emb, ref_labels)
            self.Positive = self.PositiveMiner(embedding, labels, ref_emb, ref_labels, type_of_triplets)

    def match_matrix(self, labels, ref_labels):

        if ref_labels is None:
            ref_labels = labels
        labels1 = labels.unsqueeze(1)
        labels2 = ref_labels.unsqueeze(0)
        matches = (labels1 == labels2).byte()
        matches -= torch.eye(matches.size(0)).byte().to(labels.device)
        return matches

    def diff_matrix(self, labels, ref_labels):

        if ref_labels is None:
            ref_labels = labels
        labels1 = labels.unsqueeze(1)
        labels2 = ref_labels.unsqueeze(0)
        matches = (labels1 == labels2).byte()
        diffs = matches ^ 1
        return diffs

    def RandomIndicies(self, labels, ref_labels):
        ind = self.diff_matrix(labels, ref_labels)
        indices = np.zeros(shape=(labels.shape[0],), dtype=np.int64)
        for i in range(labels.shape[0]):
            x = torch.nonzero(ind[i])
            x = torch.squeeze(x)
            indices[i] = np.random.choice(a=x)
        return indices

    def FurtherNearestIndicies(self, embedding, labels, ref_emb, ref_labels):
        neg_ind = self.diff_matrix(labels, ref_labels)
        indicies = np.zeros(shape=(labels.shape[0],), dtype=np.int64)
        with torch.no_grad():
            for i in range(labels.shape[0]):
                x = torch.nonzero(neg_ind[i])
                x = torch.squeeze(x)
                neg_embedding = ref_emb[x]
                a_n_dist = torch.nn.functional.pairwise_distance(embedding[i], neg_embedding)
                _, index_FurNear = torch.topk(a_n_dist, k=1, largest=self.furthest)
                indicies[i] = x[index_FurNear]
        return indicies

    def PositiveMiner(self, embedding, labels, ref_emb, ref_labels, type_of_triplets):
        pos_ind = self.match_matrix(labels, ref_labels)
        indicies = np.zeros(shape=(labels.shape[0],), dtype=np.int64)
        with torch.no_grad():
            negative = embedding[self.Negative]
            a_n_distance = torch.nn.functional.pairwise_distance(embedding, negative)
            for i in range(len(a_n_distance)):
                x = torch.nonzero(pos_ind[i])
                x = torch.squeeze(x)
                pos_embedding = ref_emb[x]
                a_p_dist = torch.nn.functional.pairwise_distance(embedding[i], pos_embedding)
                triplet_margin = a_p_dist - a_n_distance[i]
                if type_of_triplets == 'semihard':
                    threshold_condition = (0 < triplet_margin) & (triplet_margin < self.margin)
                    if bool(threshold_condition.sum() == 0 ) :
                        triplet_margin[triplet_margin > 0] = -math.inf
                        __, index_closest = torch.topk(triplet_margin, k=1, largest=True)
                        pos_index = x[index_closest]
                    else:
                        possible_pos = x[threshold_condition]
                        pos_index = np.random.choice(a=possible_pos)
                elif type_of_triplets == 'hard':
                    threshold_condition = self.margin <= triplet_margin
                    if threshold_condition.sum() == 0:
                        __, index_closest = torch.topk(triplet_margin, k=1, largest=True)
                        pos_index = x[index_closest]
                    else:
                        possible_pos = x[threshold_condition]
                        pos_index = np.random.choice(a=possible_pos)

                indicies[i] = pos_index

            return indicies

    def __iter__(self):
        sample_list = []
        for i, j, k in zip(self.Anchor, self.Positive, self.Negative):
            sample_list.append(i)
            sample_list.append(j)
            sample_list.append(k)

        return iter(sample_list)

    def __len__(self):
        return len(self.Anchor) + len(self.Positive) + len(self.Negative)
```

**Comment by KevinMusgrave** [comment]
What version of this library are you using? I think you may be getting versions mixed up. In the old version, the default class split would result in partition 0 getting 42% MAP on epoch -1. In order to get that same result in the latest version, you have to use the MLRCSplitManager.

**Comment by alpoler** [comment]
I use version 0.9.32. I replace ClassDisjointSplitter with MRCSplitManager. You evaluate them in 128 dimension embedding and it is saved as default . Are there any extra change that I activate PCA ?

**Comment by KevinMusgrave** [comment]
Epoch -1 refers to the pretrained trunk model, which outputs embeddings of size 512.

Epoch 0 refers o the pretrained trunk + embedder, which outputs 128 by default.

**Comment by KevinMusgrave** [comment]
Yes it makes sense to use torch.no_grad(). Also when you compute the embeddings for the offline miner, you should set your model to eval mode: ```self.models["trunk"].eval()```

**Comment by alpoler** [comment]
I implement everything related with offline miner but it concludes with MAP@R = 16.5. I use about 4400 triplet instead of 268,800 triplets. I think that insufficient triplets causes low accuracy. I observed that training reached the optimum accuracy until 30 epoch. It can't find better model after that. If ı increase the number of semi-hard and hard-triplet relative to chosed negative sample , Is it increase performance ? or Are there any essential reasons of the low accuracy in this circumstances ? @KevinMusgrave  I am glad to hear your opinion . Thanks.

**Comment by KevinMusgrave** [comment]
I don't have much experience with offline miners. Maybe it'll work better if you start with a large number of triplets, and then reduce the number as training progresses.
